docs/adr/0003-use-queue-for-backend-services.md

```diff
@@ -0,0 +1,140 @@
+- AWS SQS
+  - Amazon SQS is a message queue service used by distributed applications to exchange messages through a polling model, and can be used to decouple sending and receiving components.
+- RabbitMQ
+  - RabbitMQ is a messaging broker - an intermediary for messaging. It gives your applications a common platform to send and receive messages, and your messages a safe place to live until received.
+
+## Decision Outcome
+
+Chosen option: "GCP Pub/Sub", because Monitor services already run in GCP, and it is relatively
+simple to scale. In all cases, some level of monitoring and tuning is necessary to provide sufficiently
+capacity to ensure that tasks are not lost.
+
+This will be realized in two phases:
+
+### Phase 1
+
+- Separate out the data ingestion parts of the system, and host these separately from the user-facing APIs and pages.
+- Introduce a queue between external providers and the data ingestion parts of the system.
+- Introduce an hourly cron job to act as the worker which ingests messages and performs the task (such as looking up impacted users and sending breach alerts).
+
+A cron job is easier for SRE to support for now, but is not running often enough to use for user-facing features, and isn't as easy to scale as HTTP endpoints.
+
+### Phase 2
+
+- Replace the cron job with an internal HTTP endpoint (backed by e.g. GCP Cloud Functions, or a separate app in our existing Kubernetes cluster).
+
+### A note on PostgreSQL performance
+
+Postgres is required for the normal function of the Monitor site, and so is a single point of failure.
+This proposal focuses on queueing to be able to control the flow of tasks, but there are also some
+further improvements that have been identified that would make Postgres itself more resilient to spikes in
+demand:
+
+- Use read replica(s)
+  - Right now, Monitor uses the same Postgres instance for reads and writes
+  - The GCP Cloud SQL managed service makes this pretty straightforward to set up, but Monitor will need to be modified to always send writes to the main server and reads to the replica(s).
+- Examine Postgres statistics and adjust indexes as necessary.
+
+There are likely schema changes that would improve performance as well, but we need to be careful to
+maintain backwards compatibility during migrations.
+
+## Pros and Cons of the Options
+
+### [GCP Cloud Tasks](https://cloud.google.com/tasks)
+
+- Pros:
+  - Supports option of using cloud functions vs. having to host workers in k8s
+  - Relatively simple to set up and use compared to other options
+  - Mozilla is already using this cloud vendor
+- Cons:
+  - Locked in to GCP ecosystem
+  - Hard to test locally/offline
+    - No official emulator, unofficial open-source emulators exist
+  - Might be more difficult for external users of open-source
+  - No support for strict ordering
+    - https://cloud.google.com/tasks/docs/common-pitfalls
```

Comment:
Is this a big deal? Most of the 3rd party notification systems I've used don't claim to send them in order to begin with.

Reply:
Not for Monitor, no.